Homework_11/src/Circus.java

```diff
@@ -0,0 +1,42 @@
+import java.io.*;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Circus<T> {
+    private List<T> performers;
+
+    public Circus() {
+        this.performers = new ArrayList<>();
+    }
+
+    public void addPerformer(T performer) {
+        performers.add(performer);
+    }
+
+    public void removePerformer(T performer) {
+        performers.remove(performer);
+    }
+    public void removeAllPerformers() {
+        performers.clear();
+    }
+
+    public List<T> listPerformers() {
+        return performers;
+    }
+
+    public void saveToFile(String filename) {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(filename))) {
+            oos.writeObject(performers);
+        } catch (IOException e) {
+            System.out.println("Error saving to file: " + e.getMessage());
+        }
+    }
+
+    public void loadFromFile(String filename) {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(filename))) {
+            performers = (List<T>) ois.readObject();
```

Comment:
This can be the way to do it, although it has some disadvantages

1) The source file must contain the serialized List of objects with `T` type (so you're forced to save into a file **only** by using your implementation of `saveToFile(String filename)` method, otherwise someone who loads from file can face either Deserialization or Casting issues
2) No one will be able to prepare the file manually as the "external dataset" or export the data from another source (Database for ex.) to feed into your program - for this they'll need another script/program that transforms from their data format to the serialized format which your program understands.
3) The serialized file will be in binary format and hence hardly readable by humans (I know "make it readable by humans" was not a part of the task explicitly, but the task said "Завантажте ваші вихідні файли разом із прикладами результатів роботи (виведення консолі та/або файлів)" which ideally meant they should be human-readable
4) When loading from the file, you have no control over the particular objects from the List: what if you need to add some validation here as well **during** the loading process, so fields of each imported object can be validated? (let's say someone has saved the file incorrectly or some of the objects from the file contain invalid values - then you'll construct a new `List<T>` and forcefully push them inside anyway

Suggestions here:
1) Keep it simple until otherwise required: when saving to file -> iterate over the `performers` list and for each element, save its representation as a list of property values separated by some separator:

```text
John, Clown, 32
Henriette, Fire eater, 28
Juseppe, Gymnast, 35
```

2) When reading from a file, retrieve info line by line, and for each line extract the values, and for each value from the line make sure it has a correct type + is valid in our context and then create a new CircusPerformer object in place and set its fields using the constructor method call -> then add a CircusPerformed object to `performers` list

**Encouraging note**
Overall, the way you did it is sometimes used in real applications (when objects have to be serialized properly into a specific binary format like ProtoBuf or Avro) to maintain efficiency when storing this data and transferring through the network on high-load applications - but for now let's say we don't want to do it :)

Reply:
Thanks for the review!
I updated to use Buffered reader and writter, but i'm not sure if it is a correct approach(
BTW as we using generic for performers, i didn't understand how to convert a string(line) from a file to an object <T> with needed fields. Would really appreciate your further assistance.